src/include/btmem.h

```diff
@@ -856,6 +857,24 @@ struct __wt_ref {
 #define	WT_REF_SPLIT	 7		/* Parent page split (WT_REF dead) */
 	volatile uint32_t state;	/* Page state */
 
+	struct {
+		WT_SESSION_IMPL *session;
+		const char *name;
+		const char *file;
+		int line;
+		uint32_t state;
+	} hist[3];
+	int histoff;
+#define	WT_REF_SETSTATE(ref, s) do {					\
+	ref->hist[ref->histoff].session = session;			\
+	ref->hist[ref->histoff].name = session->name;			\
+	ref->hist[ref->histoff].file = __FILE__;			\
+	ref->hist[ref->histoff].line = __LINE__;			\
+	ref->hist[ref->histoff].state = s;				\
+	ref->histoff = (ref->histoff + 1) % (int)WT_ELEMENTS(ref->hist);\
+	WT_PUBLISH(ref->state, s);					\
+} while (0)
+
```

Comment:
I've done something like this a few times as well -- maybe we could #ifdef this on something and leave it in place? We could probably crank down the size by a bit, pass session as an argument (if session & session.name are sufficiently useful to keep permanently).

Reply:
I'll add my hat to the ring that this seems really useful, probably under diagnostic.